File: 450 Questions for DSA/Strings/57_Permutations_of_String.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
	public:

        string reverse(string S, int start, int end){
            while(start<end){
                swap(&S[start], &S[end]);
                start++;
                end--;
            }
            return S;
        }
        void swap (char* a, char* b)  
        {  
            char t = *a;  
            *a = *b;  
            *b = t;  
        }  

        int findCeil(string S, char first_character, int start, int end){
            // cout<<"String inside findceil = "<<S<<endl;
            int ceil_index = start;
            for(int i = start+1; i<=end; i++){
                if(S[i]>first_character && S[i] < S[ceil_index]){
                    ceil_index = i;
                }
            }
            return ceil_index;
        }

		vector<string>find_permutation(string S)
		{
            vector<string>ans_string;
            int size = S.length();
            sort(S.begin(), S.end());
		    bool isCompleted = false;
            while(!isCompleted){
                ans_string.push_back(S);
        
                int i;
                char first_character;
                for(i = size-2; i>=0; --i){
                    if(S[i]<S[i+1])
                        break;
                }
                if(i == -1)
                    isCompleted = true;
                
                else{
                    
                    first_character = S[i];
                    int ceil_index = findCeil(S, first_character, i+1, size-1);
                    swap(&S[i], &S[ceil_index]);

                    S = reverse(S, i+1, size-1);
                }
            }
            return ans_string;
        }
};

int main(){
    int t;
    cin>>t;
    while(t--){
        string S;
        cin>>S;
        Solution ob;
        vector<string> ans = ob.find_permutation(S);
        for(auto i: ans){
            cout<<i<<" ";
        }
        cout<<"\n";
    }
    return 0;
}

File: 450 Questions for DSA/Strings/57_Permutations of String.py
# class Solution:

#     def toString(self, List):
#         return ''.join(List)

#     def swap(self, S,pos1,pos2):
#         t = S[pos1]
#         S[pos1] = S[pos2]
#         S[pos2] = t

#     def find_permutation_helper(self,S,l,r):
#         # print(S)
#         if l == r:
#             print(self.toString(S))
#         for i in range(l,r+1):
#             self.swap(S, l, i)
#             self.find_permutation_helper(S,l+1,r)
#             self.swap(S, l, i)
        

#     def find_permutation(self, S):
#         self.find_permutation_helper(S,0,len(S)-1)


class Solution:

    ans = []
    def toString(self, List):
        return ''.join(List)

    def swap(self, S,pos1,pos2):
        t = S[pos1]
        S[pos1] = S[pos2]
        S[pos2] = t

    def find_permutation_helper(self,S,l,r):
        # print(S)
        if l == r:
            self.ans.append((self.toString(S)))
        for i in range(l,r+1):
            self.swap(S, l, i)
            self.find_permutation_helper(S,l+1,r)
            self.swap(S, l, i)
        

    def find_permutation(self, S):
        # S.sort()
        self.ans = []
        self.find_permutation_helper(list(S),0,len(S)-1)
        self.ans.sort()
        return self.ans

sol = Solution()
print(sol.find_permutation(list('ABC')))